src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -132,7 +131,7 @@ public class Task : IAsyncResult, IDisposable
         // the completion event which will be set when the Future class calls Finish().
         // But the event would now be signalled if Cancel() is called
 
-        internal object? m_stateObject; // A state object that can be optionally supplied, passed to action.
+        private protected object? m_stateObject; // A state object that can be optionally supplied, passed to action.
         internal TaskScheduler? m_taskScheduler; // The task scheduler this task runs under.
```

Comment:
Could `TaskScheduler? m_taskScheduler` be moved to `ContingentProperties`? 

Is uncommon that it's set?

Reply:
> Could TaskScheduler? m_taskScheduler be moved to ContingentProperties? Is uncommon that it's set?

I've toyed with that multiple times but I don't think we should.

It's certainly uncommon in some situations, and it's obviously unused on every async state machine, but in certain libraries it's used all over the place.  Orleans, for example, relies heavily on TaskScheduler.  And if you do have one set when you invoke an async method, every continuation is going to schedule a task to that scheduler and will have its scheduler field set appropriately, so it would make async methods invoked with a non-default scheduler significantly more expensive.

The main thing I've instead looked at is whether there's a way we could move it up off the base type.  The 99.9% case is that it's provided via factory methods, e.g. TaskFactory.StartNew, Task.ContinueWith, etc.  In some of those cases, there's already a Task-derived type that's returned, and in others (e.g. StartNew) there could be.  In such cases, the field could be moved up to those derived types and just abstracted via a base virtual Scheduler property, such that other derives types (like the state machine box) wouldn't need to pay for it.  The sticking point is Task's ctors and Task.Start, as then you can have the base type with a scheduler.  If we found a good way to make only that case more expensive and async methods / TCS tasks less expensive, I'd go for it.

I plan to have another go at it.

Reply:
Something else to factor in; is also similar with Cancellation as I've seen a huge improvement in Cancellations being plumbed though properly in the last few years; however that lives on `ContingentProperties` so that expansion will be much more common now? (though I haven't really seen it showing up as anything significant in traces that I've looked at, just source code)

Reply:
> however that lives on ContingentProperties so that expansion will be much more common now?

That will only affect tasks where the CancellationToken is provided to the task directly (e.g. Task.Run(..., token), Task.ContinueWith(..., token, ...) ) or tasks that end in a Canceled state. Note in particular that the async state machine box type isn't in the former category, so it would only have its contingent properties expanded for cancellation if it ended in a canceled state, which also means it just caught an OperationCanceledException that had been thrown.

Reply:
Ah that's why its never showed up 🙂